alternate clock configuration for the stm32 ospi driver

FRASTM opened the pull request:
For stm32 MCUs, this PR is configuring the clock source for the ospi peripheral based on the definition of the DTS.

The clock source of the Octospi is set by a bit field in one independent clock configuration register of the  RCC.
In the clock control driver, the alternate clock source of the peripheral is defined by the macro OCTOSPI_SEL() or OSPI_SEL() depending on the serie. This is valid for any stm32 device with octospi peripheral. 
The DTS of the board gives the clock source to be applied like: 
-   stm32l562e_dk
-   b_u585i_iot02a


Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 ospi driver configures peripheral clock

The clock of the octospi peripheral is directly defined
by the DTS and configured by the clock_control_on function.
No specific stm32cube function is required then.
The clock control is taking this clock source to calculate
the clock rate.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32u5 octospi clock source

The definition of the clock source for the 2 octospi
instances is given by the DTS node.
The default value selects the sysclk (not pclk)
for the alternate clock control.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1159,10 +1159,7 @@ static int flash_stm32_ospi_init(const struct device *dev)
-	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
-
-	/* Clock configuration */
+	/* Clock configuration, including the alternate clock for peripheral */
 	if (clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
 			     (clock_control_subsys_t) &dev_cfg->pclken) != 0) {
```
You miss to add some code that will query the additional clock cells. Cf https://github.com/zephyrproject-rtos/zephyr/pull/45508 for instance.

FRASTM replied:
done

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1773,12 +1788,13 @@ static void flash_stm32_ospi_irq_config_func(const struct device *dev);
 
 PINCTRL_DT_DEFINE(STM32_OSPI_NODE);
 
+static const struct stm32_pclken pclken_id[] =
+		STM32_DT_INST_CLOCKS(0);
```
This should fit in one line
```suggestion
static const struct stm32_pclken pclken_id[] = STM32_DT_INST_CLOCKS(0);
```

erwango reviewed (changes requested):
Otherwise looks good

FRASTM commented:
refine with the new  STM32_DT_CLOCKS macro from  https://github.com/zephyrproject-rtos/zephyr/pull/46842

erwango reviewed (dismissed):
LGTM, but requires https://github.com/zephyrproject-rtos/zephyr/pull/46842 to be merged first

FRASTM commented:
rebase after merge of https://github.com/zephyrproject-rtos/zephyr/pull/46842

erwango reviewed (changes requested):
One request on the setting location, otherwise LGTM.

FRASTM committed:
dts: arm: stm32l5 octospi clock source

The definition of the octospi clock source is given
by the DTS node. The default value selects the sysclk
(not pclk) for the alternate clock control.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi:
```diff
@@ -109,6 +109,9 @@
 &octospi2 {
+	/delete-property/ clocks;
+	clocks = <&rcc STM32_CLOCK_BUS_AHB2_2 0x00000100>,
+		<&rcc STM32_SRC_SYSCLK OCTOSPI_SEL(0)>;
```
moved from board DTS to soc DTS of the stm32u5

FRASTM commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -128,6 +128,9 @@
 &octospi1 {
+	/delete-property/ clocks;
+	clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00000100>,
+		<&rcc STM32_SRC_SYSCLK OSPI_SEL(0)>;
```
moved from board DTS to soc DTS of the stm32l5
